DashKit/DashPumpManager.swift

```diff
@@ -940,17 +980,170 @@ public class DashPumpManager: PumpManager {
+        
+        var calendar = Calendar(identifier: .gregorian)
+        calendar.timeZone = state.timeZone
+        
+        self.mutateState { (state) in
+            switch pendingCommand {
+            case .program(let program, let commandDate):
+                let scheduledSegmentAtCommandTime = state.basalProgram.currentRate(using: calendar, at: commandDate)
+
+                if let dose = program.unfinalizedDose(at: commandDate, withCertainty: .uncertain) {
+                    switch dose.doseType {
+                    case .bolus:
+                        if dose.isFinished(at: dateGenerator()) {
+                            state.finishedDoses.append(dose)
+                        } else {
+                            state.unfinalizedBolus = dose
+                        }
+                    case .tempBasal:
+                        // Assume a high temp succeeded, but low temp failed
+                        let rate = dose.programmedRate ?? dose.rate
+                        if rate > scheduledSegmentAtCommandTime.basalRateUnitsPerHour {
+                            if dose.isFinished(at: dateGenerator()) {
+                                state.finishedDoses.append(dose)
+                            } else {
+                                state.unfinalizedTempBasal = dose
+                            }
+                        }
+                    case .resume:
+                        state.finishedDoses.append(dose)
+                    case .suspend:
+                        break // start program is never a suspend
+                    }
+                }
+            case .stopProgram(let stopProgram, let commandDate):
+                let scheduledSegmentAtCommandTime = state.basalProgram.currentRate(using: calendar, at: commandDate)
+                
+                // All stop programs result in reduced delivery, except for stopping a low temp, so we assume all stop
+                // commands failed, except for low temp
+                var tempBasalCancel = false
+
+                switch stopProgram {
+                case .tempBasal:
+                    tempBasalCancel = true
+                case .stopAll:
+                    tempBasalCancel = true
+                default:
+                    break
+                }
+                
+                if tempBasalCancel,
+                    let tempBasal = self.state.unfinalizedTempBasal,
+                    !tempBasal.isFinished(at: commandDate),
+                    (tempBasal.programmedRate ?? tempBasal.rate) < scheduledSegmentAtCommandTime.basalRateUnitsPerHour
+                {
+                    state.unfinalizedTempBasal?.cancel(at: commandDate)
+                }
+            }
+            state.pendingCommand = nil
+        }
+    }
+
+    public func attemptUnacknowledgedCommandRecovery() {
+        if let pendingCommand = self.state.pendingCommand {
+            podCommManager.queryAndClearUnacknowledgedCommand { (result) in
```

Comment:
Wondering if there is any concern that the pending command in `state` doesn't match the unacknowledged command on the pump? Or am I overlooking some check that these are the same command?

Reply:
The pump won't accept any new commands when there is an unacknowledged command.